### PR TITLE
Refactor of VirtualNode-controller.

### PR DESCRIPTION
**Pull request by Andreagit97**
# Description

This PR allows managing in a simpler way the NamespaceMap lifecycle. Now the controller has only one main task: 
- the controller regenerates its NamespaceMap when it is deleted.

Ref. #572 STEP 12

### CODE REVIEW

File: pkg/liqo-controller-manager/virtualNode-controller/manage_virtualnode_controller_finalizer.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package virtualnodectrl
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/klog"
```
Comment (palexster) [inline pkg/liqo-controller-manager/virtualNode-controller/manage_virtualnode_controller_finalizer.go:7]: Should be v2

File: pkg/liqo-controller-manager/virtualNode-controller/namespaceMap_creation.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
+package virtualnodectrl
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog"
```
Comment (palexster) [inline pkg/liqo-controller-manager/virtualNode-controller/namespaceMap_creation.go:9]: ditto

File: pkg/liqo-controller-manager/virtualNode-controller/namespaceMap_creation.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	ctrlutils "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	mapsv1alpha1 "github.com/liqotech/liqo/apis/virtualKubelet/v1alpha1"
+	liqoconst "github.com/liqotech/liqo/pkg/consts"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
```
Comment (palexster) [inline pkg/liqo-controller-manager/virtualNode-controller/namespaceMap_creation.go:15]: should be grouped

File: pkg/liqo-controller-manager/virtualNode-controller/suite_test.go [hunk 4]
```diff
@@ -128,6 +129,26 @@ var _ = BeforeSuite(func(done Done) {
 	k8sClient = k8sManager.GetClient()
 	Expect(k8sClient).ToNot(BeNil())
 
+	nms = &mapsv1alpha1.NamespaceMapList{}
+	technicalNamespace = &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: liqoconst.TechnicalNamespace,
+		},
+	}
+
```
Comment (palexster) [inline pkg/liqo-controller-manager/virtualNode-controller/suite_test.go:138]: Can you add some By to specify what are you testing? The test-case description is very limited

File: pkg/liqo-controller-manager/virtualNode-controller/virtualnode_deletion_logic.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+	mapsv1alpha1 "github.com/liqotech/liqo/apis/virtualKubelet/v1alpha1"
+	liqoconst "github.com/liqotech/liqo/pkg/consts"
+)
+
+func (r *VirtualNodeReconciler) removeAssociatedNamespaceMaps(ctx context.Context, n *corev1.Node) error {
+	klog.Infof("The virtual virtualNode '%s' is requested to be deleted", n.GetName())
+
+	// The deletion timestamp is automatically set on the NamespaceMaps associated with the virtual-node,
+	// it's only necessary to wait until the NamespaceMaps are deleted.
+	nms := &mapsv1alpha1.NamespaceMapList{}
```
Comment (palexster) [inline pkg/liqo-controller-manager/virtualNode-controller/virtualnode_deletion_logic.go:20]: Can you use a more explicit variable? Nm can be confused with namespace

**Review by palexster — APPROVED** [review]

File: pkg/liqo-controller-manager/virtualNode-controller/virtualnode_controller.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+	if !virtualNode.GetDeletionTimestamp().IsZero() {
+		return ctrl.Result{}, r.removeAssociatedNamespaceMaps(ctx, virtualNode)
+	}
+
+	// The virtual-node must have the cluster-id annotation
+	if _, ok := virtualNode.GetAnnotations()[liqoconst.RemoteClusterID]; !ok {
+		err := fmt.Errorf("the annotation '%s' is not found on node '%s'", liqoconst.RemoteClusterID, virtualNode.GetName())
+		klog.Error(err)
+		return ctrl.Result{}, err
+	}
```
Comment (giorio94) [inline pkg/liqo-controller-manager/virtualNode-controller/virtualnode_controller.go:71]: Should be moved above the previous function, as the annotation is used also in that case.

File: pkg/liqo-controller-manager/virtualNode-controller/namespaceMap_creation.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+	if err := r.Create(ctx, nm); err != nil {
+		klog.Errorf("%s --> Problems in NamespaceMap creation for the virtual node '%s'", err, n.GetName())
+		return err
+	}
+	klog.Infof(" Create the NamespaceMap '%s' for the virtual node '%s'", nm.GetName(), n.GetName())
+	return nil
+}
+
+// This function manages NamespaceMaps Lifecycle on the basis of NamespaceMaps' number.
+func (r *VirtualNodeReconciler) checkNamespaceMapPresence(ctx context.Context, n *corev1.Node) error {
```
Comment (giorio94) [inline pkg/liqo-controller-manager/virtualNode-controller/namespaceMap_creation.go:42]:
```suggestion
func (r *VirtualNodeReconciler) ensureNamespaceMapPresence(ctx context.Context, n *corev1.Node) error {
```

Same as before about name clarity.

**Review by giorio94 — APPROVED** [review]

File: pkg/consts/namespace_mapping.go [hunk 8]
```diff
@@ -3,8 +3,8 @@ package consts
 const (
 	// RemoteClusterID is used to obtain cluster-id from different Liqo resources.
 	RemoteClusterID = "cluster-id" // "remote.liqo.io/clusterId"
-	// MapNamespaceName is the namespace where NamespaceMap are created.
-	MapNamespaceName = "default"
+	// TechnicalNamespace is the namespace where NamespaceMap are created.
+	TechnicalNamespace = "default"
```
[the remark this answers is not in the document]
Comment (Andreagit97) [reply pkg/consts/namespace_mapping.go:7]: You are right, the NamespaceMaps should be created in the "Technical Namespace" not in the default one, ATM I will change the default namespace with the "Liqo Namespace". The NamespaceMaps will be moved in the technical namespace in a future PR. The name of the technical namespace should not be known at install time...

File: pkg/liqo-controller-manager/virtualNode-controller/manage_virtualnode_controller_finalizer.go [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrlutils "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+func (r *VirtualNodeReconciler) checkVirtualNodeFinalizerPresence(ctx context.Context, n *corev1.Node) error {
```
[the remark this answers is not in the document]
Comment (Andreagit97) [reply pkg/liqo-controller-manager/virtualNode-controller/manage_virtualnode_controller_finalizer.go:12]: you are right, thank you

File: pkg/liqo-controller-manager/virtualNode-controller/virtualnode_controller.go (hunk 6 above)
Comment (Andreagit97) [reply pkg/liqo-controller-manager/virtualNode-controller/virtualnode_controller.go:71]: you are right, thank you

**Pull request merged**